[FLINK-2730][web-dashboard] Adds cpu and memory usage graphs.

sachingoel0101 opened the pull request:
Screenshots:
![screenshot 21](https://cloud.githubusercontent.com/assets/8874261/10336674/1d1d9e60-6d18-11e5-94f8-7308098789bf.png)
![screenshot 22](https://cloud.githubusercontent.com/assets/8874261/10336675/1d2a8bf2-6d18-11e5-9dde-194097ebf103.png)
![screenshot 23](https://cloud.githubusercontent.com/assets/8874261/10336676/1d2c22b4-6d18-11e5-8ca5-0abd44349f0b.png)
![screenshot 24](https://cloud.githubusercontent.com/assets/8874261/10336677/1d3434f4-6d18-11e5-9b0a-3f95c40d90ae.png)

I would really like to get this in the release. One more step towards removing the old frontend. :)

aljoscha commented:
Wow, the screenshots look super nice. :smile:

Someone with knowledge about the new web fronted should probably review the code, though.

tillrohrmann commented:
Really cool graphs Sachin :-) Looking forward monitoring my Flink jobs with
these graphs.

On Wed, Oct 7, 2015 at 3:10 PM, Aljoscha Krettek <notifications@github.com>
wrote:

> Wow, the screenshots look super nice. [image: :smile:]
>
> Someone with knowledge about the new web fronted should probably review
> the code, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/1236#issuecomment-146191633>.
>

sachingoel0101 commented:
Thanks. :)
Although there's not much to review here. :') It's mostly a rearrangement of existing content, with a major addition in `taskmanager.ctrl.coffee`

aljoscha commented:
@iampeter, could you please have a look so that we can quickly merge this?

iampeter commented:
@aljoscha ok will do

iampeter commented:
@sachingoel0101 @aljoscha looks great, but I would make the following changes if possible:

1. Move the 'livechart' directive to a separate file - `taskmanager.dir.coffee`
2. Move the Highcharts config in `index.coffee` to either `.run()` in `index.coffee`, or the directive
3. Remove the 'No such task manager exits' text that blinks when the task manager details are loading

Thanks,
Piotr

aljoscha commented:
Thanks for the review @iampeter :smile:

sachingoel0101 commented:
@iampeter , thanks for the review. I've addressed all your comments.
I removed the `No such task ...` thing entirely. It doesn't have a use case anyways. We don't expect the user to break things on the server here. If the task manager doesn't exist, it is handled gracefully by the backend and an empty array is returned.

sachingoel0101 commented:
@aljoscha FYI I've added a new Job Manager message which returns the metrics only for one particular instance to minimize unnecessary traffic for a single taskmanager web page. That okay?

sachingoel0101 commented:
I think I broke something while addressing Piotr's concerns. Will debug and push the fixed version again.

mxm commented:
The screenshots look great. I'll check out your work now :)

uce commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagersHandler.java:
```diff
@@ -39,6 +42,7 @@
 
 	private final FiniteDuration timeout;
 
+	private static final String taskManagerIDKey = "taskmanagerid";
```
This is referencing your addition in WebRuntimeMonitor. You can use this constant there as well. I would furthermore use the constant field style a la "TASKMANAGER_ID_KEY".

sachingoel0101 replied:
Pushed a fix.

mxm commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -536,6 +536,11 @@ class JobManager(
+      sender ! decorateMessage(
+        TaskManagerInstance(instanceManager.getRegisteredInstanceById(instanceID))
+      )
+
```
Wouldn't it be better to return an InstanceNotFound message instead of a null instance?

sachingoel0101 replied:
I think using an Option makes more sense here.

mxm commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagersHandler.java:
```diff
@@ -54,10 +58,22 @@ public String handleRequest(Map<String, String> params) throws Exception {
+					InstanceID instanceID = new InstanceID(StringUtils.hexStringToByte(params.get(TASK_MANAGER_ID_KEY)));
+					Future<Object> future = jobManager.ask(new JobManagerMessages.RequestTaskManagerInstance(instanceID), timeout);
+					TaskManagerInstance instance = (TaskManagerInstance) Await.result(future, timeout);
+					if (instance.instance() != null) {
```
Not sure about the null check. Could check for InstanceNotFound message instead.

sachingoel0101 replied:
Will fix this.

mxm commented:
Tried out your changes it everything worked without quirks. The initial look is a bit odd when the charts haven't loaded yet but this is nothing that should prevent us from merging the PR. I've also made a comment inline.

+1 for merging

sachingoel0101 commented:
The initial look is a bit weird, I agree. There is a potential fix, by fetching the usage upto the point when someone clicks on the `metrics` tab, however, that can be quite large if the cluster has been running for a long time. This would require storing the metrics history on the job manager, which will unnecessarily waste memory.

sachingoel0101 commented:
@mxm I modified the code to remove null values. It should be mergeable now.

mxm commented:
Looks good to merge now.

sachingoel0101 commented:
Build passes successfully. Looking forward to seeing this merged. :)

mxm commented:
Merging this after a final test...

mxm commented on flink-runtime-web/web-dashboard/bower.json:
```diff
@@ -23,7 +23,8 @@
     "moment-duration-format": "~1.3.0",
-    "qtip2": "~2.2.1"
+    "qtip2": "~2.2.1",
+    "highcharts-release": "~4.1.8"
```
This library's license is incompatible with the Apache License. Unfortunately, nobody noticed this. We will have to revert the merged changes...

sachingoel0101 replied:
I apologize. It should've been my responsibility primarily.
I think I might be able to achieve the same graphs and live updates using `d3`, which I assume is compatible. Working on it right now.

StephanEwen replied:
Thanks! For now, we'll try to make a release without that page. We can add that back for 0.10.1 and 1.0

sachingoel0101 replied:
Okay. I can a open a PR to minimally remove this functionality unless the commit needs to be completely reverted.

mxm replied:
I'm already working on a change to remove only the charts. Apparently, `d3` would work for us because it is BSD-lincensed.

sachingoel0101 replied:
Great. I'll start working on d3 based charts after you push your commit.
